Refinement types and unification for symbolic shape inference

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62682
* #62582
* **#61776**

Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

migeed-z committed:
Refinement types and unification for symbolic shape inference

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"




[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"




[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"




[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"




[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"




[ghstack-poisoned]

vadimkantorov commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +511,148 @@ def type_check_node(self, n: Node):
+class Refine:
+    """
+    Symbolic shape inference.
+    Generates constraitns over type variables.
```
typo: constraitns

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"




[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -6,8 +6,16 @@
+def conv3x3(in_planes, out_planes, stride=1, groups=1, dilation=1):
+    """3x3 convolution with padding"""
+    return torch.nn.Conv2d(in_planes, out_planes, kernel_size=3, stride=stride,
+                           padding=dilation, groups=groups, bias=False, dilation=dilation)
```
Is this function used?

migeed-z replied:
yea you're right, I got rid of it

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +516,148 @@ def type_check_node(self, n: Node):
+    assert isinstance(n.args[0], Node)
+    arg_type = n.args[0].type
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        return [Equality(arg_type.__args__[0], n.type.__args__[0])]
```
So this is just saying that the batch size at input needs to match the batch size at output?

Can we encode e.g. that the channel dimensions at input/output need to match the parameter value? Or is that not necessary?

migeed-z replied:
We already have this in the initial pass

```
        w_in = arg_type.__args__[3]
        h_in = arg_type.__args__[2]
        h_out = calculate_out_dimension(h_in, module_instance, 0)
        w_out = calculate_out_dimension(w_in, module_instance, 1)
        new_type = TensorType((arg_type.__args__[0], module_instance.out_channels, h_out, w_out))
```

So I think we already have that part but maybe we consider moving it here at some point so we can do one pass instead of two passes. I haven't worked out what all the constraints would look like in that case though, so I'll think about it.

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +516,148 @@ def type_check_node(self, n: Node):
+    if isinstance(arg_type, TensorType) and isinstance(n.type, TensorType):
+        args1 = arg_type.__args__
+        args2 = n.type.__args__
+        return [Equality(args1[i], args2[i]) for i in range(len(args1))]
```
For add, isn't this not true in the presence of broadcasting?

migeed-z replied:
assuming we run the type inferencer first, broadcasting would have been applied to the types by this point, but maybe if types are not present we don't know if broadcasting happens? I think tomorrow we can discuss this point

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +516,148 @@ def type_check_node(self, n: Node):
+    def __init__(self, traced):
+        self.constraints = []
+        self.traced = traced
+        self.curr_symbol = self.symbol_gen()
```
nit: you can use `self.symbol_iter = itertools.count(start=0, step=1)` and then use `next(self.symbol_iter)` to retrieve the value and advance the iterator

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +516,148 @@ def type_check_node(self, n: Node):
+                    new_args.append(a)
+            return TensorType(tuple(new_args))
+        else:
+            return typ
```
style: Can we make a general recursive map function similar to [map_aggregate](https://github.com/pytorch/pytorch/blob/c30048fccfc5914cb3924d9b2905a5e13b3e0b52/torch/fx/node.py#L537) for doing this type of stuff? That way the code for traversing type expressions and yielding a new one will all be in one place and we can update it once and all of the call-sites will have the new behavior

jamesr66a commented on torch/fx/experimental/unify_refinements.py:
```diff
@@ -0,0 +1,83 @@
+    till reaching a fixed point. If the input and output graph
+    are the same, we converge.
+    """
+    old_graph = deepcopy(graph)
```
Any way we could do this in-place? I think this is already going to be very high complexity and having to copy over the entire graph every time may make it even higher

migeed-z replied:
I am doing this in place now with a dictionary. Does that seem fine or can there be a better way to do this?

jamesr66a reviewed (commented):
Looks cool! Just did a quick pass over it, will look deeper into the algorithmic aspects and play around with it tomorrow or wednesday

jamesr66a reviewed (commented):
BTW can you add a commit message on the PR?

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

eellison commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +522,169 @@ def type_check_node(self, n: Node):
+            r = []
+            for x, y, z in zip(args1, args2, args3):
+                if args1 == args2:
+                    r.append(Equality(x, z))
```
i think this isnt handling broadcasting

migeed-z replied:
I updated this part. I think that if we symbolically determine if certain elements of the arguments to addition are equal, then we know that broadcasting did not happen. But this required two passes. One to determine the information about the arguments and another to relate the arguments to the node type.

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

eellison commented on test/fx/test_gradual_type.py:
```diff
@@ -819,6 +823,53 @@ def test_resnet50(self):
+        # apply shape inference to graph and check
+        # that the batch size is equal across all layers
+        infer_symbolic_types(gm_static)
+        infer_symbolic_types(gm_static)
```
you mentioned elsewhere that you need to run it twice - should the be invoked internally as part of `infer_symbolic_types` instead of being required at the callsite ?

migeed-z replied:
yeah I will do that. The tests are calling infer_symbolic_types twice. I'll change that to a single call.

eellison commented on test/fx/test_gradual_type.py:
```diff
@@ -819,6 +823,53 @@ def test_resnet50(self):
+        infer_symbolic_types(gm_static)
+
+
+        batch_sizes = []
```
nit: not really a reason to not use set here from the start

eellison commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
+
+
+@register_refinement_rule(Conv2d)
+@register_refinement_rule(torch.nn.Linear)
```
fyi: Linear has broadcasting behavior because it dispatches to matmul https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/native/Linear.cpp.html#37

Not sure if `Linear` the module is different but:

```
>>> r = torch.rand
>>> torch._C._nn.linear(r(4), r(8, 4), r(8)).size()
torch.Size([8])
```

migeed-z replied:
I will check this! I went based on the documentation here https://pytorch.org/docs/stable/generated/torch.nn.Linear.html which talked about the input and output shapes.

eellison replied:
As you've written it I think is fine for `nn.Linear` the Module..

eellison commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
+            args3 = n.type.__args__
+            r = []
+            for x, y, z in zip(args1, args2, args3):
+                if x == y:
```
I think it would be easy enough to add a check for the `1`  behavior of broadcasting

migeed-z replied:
this part is handled in the type checker already! the definitions are here https://github.com/pytorch/pytorch/blob/faf82478fec1d777f1120ee2d1e06f01a1bf3325/torch/fx/experimental/graph_gradual_typechecker.py#L45

by the time we reach the symbolic inference part, we already did the best we can with any concrete types we had earlier.

eellison replied:
i guess the case i'm talking about is if one value is `1` the output will be the symbolic shape of the other

migeed-z replied:
I added a test case for this here https://github.com/pytorch/pytorch/pull/61776/files#diff-8ccda303dbd4c687535083b21b946c204650cb45561330e48b56a2299ab92752R922

is that along the lines of what you're referring to?

eellison commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
+    start_dim = 1
+    end_dim = -1
+
+    if len(n.args) > 1:
```
would it be easier to assume that the arguments are normalized before this ? i dont know if this handles things like kwargs

eellison commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
+            eq_const.append(Equality(t1, t2))
+
+        for t1, t2 in zip(n.type.__args__[start_dim:end_dim], n.args[0].type.__args__[start_dim:end_dim]):
+            eq_const.append(Equality(t1, t2))
```
Is this constraint right ? the output will not have the same number of dimensions as the input. 

```
>>> x = torch.rand([2, 3, 4, 5])
>>> torch.flatten(x, start_dim=1, end_dim=3).size()
torch.Size([2, 60])
```
dims 1 and 2 (iterating from `[1:3]`) will not have the same dimensions for the output and input

migeed-z replied:
This part should be fixed now

eellison reviewed (commented):
looks good, a few comments

migeed-z commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +522,155 @@ def type_check_node(self, n: Node):
+            args2 = arg_type2.__args__
+            args3 = n.type.__args__
+            r = []
+            for x, y, z in zip(args1, args2, args3):
```
since we type-checked the input first, if both operants and n.type are all tensors, then we know that the length must be equal. I think I should eventually put everything in one function instead of calling symbolic inference twice and calling the type-checker once.

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

eellison commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -500,3 +531,156 @@ def type_check_node(self, n: Node):
+            eq_const.append(Equality(t1, t2))
+
+        for t1, t2 in zip(n.type.__args__[start_dim:end_dim], n.args[0].type.__args__[start_dim:end_dim]):
+            eq_const.append(Equality(t1, t2))
```
i think the types from start_dim:end_dim are different between input/output

migeed-z replied:
I thought we had 3 parts to this. So from 0 to start and from start to end. Then start to end would be flattened but 0 to start and start to end would be the same dimensions?

eellison reviewed (commented):
lgtm, just a couple comments

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).

jamesr66a reviewed (approved):
Looks good to me, cc @eellison for any final comments

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z committed:
Update on "Refinement types and unification for symbolic shape inference"


Differential Revision: [D29772537](https://our.internmc.facebook.com/intern/diff/D29772537)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29772537).
